### PR TITLE
implement system contracts for mint

### LINKED ISSUE #8577: MINT_UNITS (MintPrecompile)

### LINKED ISSUE #8578: MINT_NFTS (MintPrecompile)

### PR DESCRIPTION
**Pull request by dikel**
**Description**:
Implementation for HTS `MINT_UNITS` and `MINT_NFTS` functions

1. Modified `MintTranslator` to use `DispatchForResponseCodeHtsCall ` instead of `MintCall`
2. Added `newTotalSupply` to `TokenMintRecordBuilder`
3. Created MintXTest for both fungibles and NFTs
4. Written unit tests

**Related issue(s)**:

Fixes #8577 
Fixes #8578

### CODE REVIEW

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/mint/MintTranslator.java [hunk 1]
```diff
@@ -58,18 +63,54 @@ public boolean matches(@NonNull final HtsCallAttempt attempt) {
...
             return token.tokenType() == TokenType.FUNGIBLE_COMMON
-                    ? new FungibleMintCall(attempt.enhancement())
-                    : new NonFungibleMintCall(attempt.enhancement());
+                    ? fungibleCallFrom(attempt.senderAddress(), call.get(0), amount, attempt)
+                    : nonFungibleCallFrom(attempt.senderAddress(), call.get(0), call.get(2), attempt);
         }
     }
+
+    private FungibleMintCall fungibleCallFrom(
+            @NonNull final org.hyperledger.besu.datatypes.Address sender,
```
Comment (agadzhalov) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/mint/MintTranslator.java:85]: we can remove `org.hyperledger.besu.datatypes.` and leave it just `Address` as we have the import on top

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/mint/MintTranslator.java [hunk 2]
```diff
@@ -58,18 +63,54 @@ public boolean matches(@NonNull final HtsCallAttempt attempt) {
...
+                attempt.enhancement(),
+                amount,
+                ConversionUtils.asTokenId(token),
+                attempt.defaultVerificationStrategy(),
+                sender,
+                attempt.addressIdConverter());
+    }
+
+    private NonFungibleMintCall nonFungibleCallFrom(
+            @NonNull final org.hyperledger.besu.datatypes.Address sender,
```
Comment (agadzhalov) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/mint/MintTranslator.java:99]: again we can remove `org.hyperledger.besu.datatypes.` and leave it just `Address` as we have the import on top

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/mint/NonFungibleMintCall.java [hunk 3]
```diff
@@ -16,20 +16,82 @@
...
+            return gasOnly(revertResult(recordBuilder.status(), 0L));
+        } else {
+            final var encodedOutput = MintTranslator.MINT
+                    .getOutputs()
+                    .encodeElements(BigInteger.valueOf(ResponseCodeEnum.SUCCESS.protoOrdinal()));
+            return gasOnly(successResult(encodedOutput, recordBuilder.getNewTotalSupply()));
+        }
+    }
+
+    private TransactionBody syntheticMintUnits(@NonNull final TokenID tokenId, final List<Bytes> metadata) {
```
Comment (agadzhalov) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/mint/NonFungibleMintCall.java:89]: we should rename `syntheticMintUnits` to smth like `syntheticMintNfts`

**Review by tinker-michaelj — CHANGES_REQUESTED** [review]
Overall looks great; but I think we should be able to implement the `MintTranslator` to return an instance of `DispatchForResponseCodeHtsCall`  (as [here](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/associations/AssociationsTranslator.java#L68)) instead of having more one-off implementations of `HtsCall`.

File: hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/hts/mint/MintTranslatorTest.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        final var matches = subject.matches(attempt);
+        assertThat(matches).isTrue();
+    }
+
+    @Test
+    void matchesMintV2Test() {
+        given(attempt.selector()).willReturn(MintTranslator.MINT_V2.selector());
+        final var matches = subject.matches(attempt);
+        assertThat(matches).isTrue();
+    }
```
Comment (lukelee-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/hts/mint/MintTranslatorTest.java:54]: Can the `callFrom` method be tested?

Comment (dikel) [reply hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/hts/mint/MintTranslatorTest.java:57]: It is tested in `HtsCallAttemptTest`

**Review by tinker-michaelj — APPROVED** [review]
This looks great, tyvm @dikel !

**Pull request merged**